### PR TITLE
Lock on write proc

### LINKED ISSUE #1570: fulltext index creation isn't replicated

### PR DESCRIPTION
**Pull request by swilly22**
Resolves #1570

### CODE REVIEW

File: src/schema/schema.c [hunk 1]
```diff
@@ -85,34 +85,47 @@ int Schema_AddIndex(Index **idx, Schema *s, const char *field, IndexType type) {
 	return INDEX_OK;
 }
 
-int Schema_RemoveIndex(Schema *s, const char *field, IndexType type) {
+static int _schema_RemoveExactMatchIndex(Schema *s, const char *field) {
```
Comment (jeffreylovitz) [inline src/schema/schema.c:88]: Nit - `_Schema_RemoveExactMatchIndex`

File: src/schema/schema.c [hunk 2]
```diff
@@ -85,34 +85,47 @@ int Schema_AddIndex(Index **idx, Schema *s, const char *field, IndexType type) {
...
-			Index_Free(idx);
-			s->index = NULL;
-		}
+		s->index = NULL;
 	}
 
 	return INDEX_OK;
 }
 
+static int _schema_RemoveFullTextIndex(Schema *s) {
```
Comment (jeffreylovitz) [inline src/schema/schema.c:107]: Nit - _`Schema_RemoveFullTextIndex`

File: src/graph/graphcontext.c [hunk 3]
```diff
@@ -375,25 +375,35 @@ int GraphContext_AddIndex(Index **idx, GraphContext *gc, const char *label,
...
 
 	// Retrieve the schema for this label
-	Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
 	int res = INDEX_FAIL;
-	if(s != NULL) res = Schema_RemoveIndex(s, field, type);
-	ResultSet *result_set = QueryCtx_GetResultSet();
-	ResultSet_IndexDeleted(result_set, res);
+	Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
+
+	if(s != NULL) {
```
Comment (jeffreylovitz) [inline src/graph/graphcontext.c:398]: Nice catch!

File: src/procedures/proc_fulltext_create_index.c [hunk 4]
```diff
@@ -23,33 +23,29 @@ ProcedureResult Proc_FulltextCreateNodeIdxInvoke(ProcedureCtx *ctx,
...
-	}
-
-	// Introduce fields to index.
+	// introduce fields to index
 	for(int i = 0; i < fields_count; i++) {
 		const char *field = fields[i].stringval;
-		// It's OK to add existing field.
-		Index_AddField(idx, field);
+		if(GraphContext_AddIndex(&idx, gc, label, field, IDX_FULLTEXT) == INDEX_OK) {
+			res = INDEX_OK;
```
Comment (jeffreylovitz) [inline src/procedures/proc_fulltext_create_index.c:43]:
```
res = GraphContext_AddIndex(&idx, gc, label, field, IDX_FULLTEXT);
if(res != INDEX_OK) break;
```
FWIW the older approach had the benefit of resolving the schema only once, but I don't mind this change if you think it's a better abstraction.

**Review by jeffreylovitz — CHANGES_REQUESTED** [review]
:+1: Looks good! A few small comments.

File: src/procedures/proc_fulltext_create_index.c (hunk 4 above)
Comment (swilly22) [reply src/procedures/proc_fulltext_create_index.c:43]: not only abstraction, but this way will make sure the statistics are updated, we can't break here as there might be several fields being added out of which some might already exists. we want to construct only if the user presented a new field

**Pull request merged**